tracer/src/Datadog.Trace/IAST/Analyzers/HardcodedSecretsAnalyzer.cs

```diff
@@ -0,0 +1,214 @@
+// </copyright>
+
+#nullable enable
+
+using System;
+using System.Collections.Generic;
+using System.Runtime.InteropServices;
+using System.Text;
+using System.Text.RegularExpressions;
+using System.Threading;
+using System.Threading.Tasks;
+using Datadog.Trace.ClrProfiler;
+using Datadog.Trace.Configuration;
+using Datadog.Trace.Logging;
+using Datadog.Trace.Vendors.Serilog.Core;
+
+namespace Datadog.Trace.Iast.Analyzers;
+
+internal class HardcodedSecretsAnalyzer
+{
+    private const int UserStringsArraySize = 100;
+    private static readonly IDatadogLogger Log = DatadogLogging.GetLoggerFor<HardcodedSecretsAnalyzer>();
+    private static HardcodedSecretsAnalyzer? _instance = null;
+
+    private static bool _started = false;
+
+    private static List<SecretRegex>? _secretRules = null;
+
+    private static ManualResetEventSlim _waitEvent = new ManualResetEventSlim(false);
+
+    public HardcodedSecretsAnalyzer()
+    {
+        Log.Debug("HardcodedSecretsAnalyzer -> Init");
+        LifetimeManager.Instance.AddShutdownTask(RunShutdown);
+        _started = true;
+        Task.Run(() => PoolingThread());
+    }
+
+    private static void PoolingThread()
+    {
+        try
+        {
+            Log.Debug("HardcodedSecretsAnalyzer polling thread -> Started");
+            while (_started)
+            {
+                var userStrings = new UserStringInterop[UserStringsArraySize];
+                int userStringLen = NativeMethods.GetUserStrings(userStrings.Length, userStrings);
+                Log.Debug("HardcodedSecretsAnalyzer polling thread -> Retrieved {UserStringLen} strings", userStringLen.ToString());
+                if (userStringLen > 0 && Tracer.Instance.Settings.IsIntegrationEnabled(IntegrationId.HardcodedSecret))
+                {
+                    for (int x = 0; x < userStringLen; x++)
+                    {
+                        var value = Marshal.PtrToStringUni(userStrings[x].Value);
+                        var match = CheckSecret(value!);
+                        if (!string.IsNullOrEmpty(match))
+                        {
+                            var location = Marshal.PtrToStringUni(userStrings[x].Location);
+                            IastModule.OnHardcodedSecret(new Vulnerability(
+                                VulnerabilityTypeName.HardcodedSecret,
+                                (VulnerabilityTypeName.HardcodedSecret + ":" + location!).GetStaticHashCode(),
+                                new Location(location!),
+                                new Evidence(match!),
+                                IntegrationId.HardcodedSecret));
+                        }
+                    }
+
+                    if (userStringLen == userStrings.Length)
+                    {
+                        continue; // Skip wait time if array came full
+                    }
+                }
+
+                _waitEvent.Wait(5_000);
```

Comment:
On shutdown, this could add a big delay (5s), which will be unacceptable in some scenarios, and could mean we don't flush traces/telemetry. We ideally should immediately exit if we're waiting when the app exits. 

I suggest you use a pattern we have elsewhere (e.g. see `DiscoveryService.FetchConfigurationLoopAsync()` or `TelemetryController.PushHeartbeatLoopAsync()`). These use

- An `async` loop
- A `TaskCompletionSource<bool>` to handle process exit
- `await Task.WhenAny(Task.Delay(5_000), _processExit.Task).ConfigureAwait(false);` to allow immediately breaking out of the wait on process exit

Reply:
That's why there's this ` _waitEvent.Set();` at shutdown so it cancels the wait in that case. But I'll have a look at the examples you say.

Reply:
The `_waitEvent` is signaled on shutdown